### PR TITLE
Process Ivars, Create custom types for ObjC Classes

### LINKED ISSUE #17: Define types for the `self*` argument in messages, allowing inlining of ivars

### LINKED ISSUE #27: Pointer types are not parsed by the Type Parser
This is potentially a duplicate of #8 only in the sense that #8 is so vague it could basically be "Issue: product has bugs"

---

In typestrings, Pointers to types, e.g. `^@` == "id *", `^^^^c` == "char ****" are incorrectly parsed.

### PR DESCRIPTION
**Pull request by cxnder**
Resolves #17 , Resolves #27

### CODE REVIEW

File: Core/AnalysisInfo.h [hunk 1]
```diff
@@ -80,6 +80,38 @@ struct MethodListInfo {
     bool hasDirectSelectors() const;
 };
 
+/**
+ * A description of an Objective-C ivar
```
Comment (cxnder) [inline Core/AnalysisInfo.h:84]: Punctuation

File: Core/AnalysisInfo.h [hunk 2]
```diff
@@ -80,6 +80,38 @@ struct MethodListInfo {
...
+    std::string name;
+    std::string type;
+
+    uint64_t offsetAddress {};
+    uint64_t nameAddress {};
+    uint64_t typeAddress {};
+    uint32_t size {};
+
+    /**
+     * Get the method's type as series of C-style tokens.
```
Comment (cxnder) [inline Core/AnalysisInfo.h:100]: Incorrect comment

File: Core/AnalysisInfo.h [hunk 3]
```diff
@@ -80,6 +80,38 @@ struct MethodListInfo {
...
+     */
+    std::string decodedTypeToken() const;
+};
+
+/**
+ * A description of an Objective-C ivar list
```
Comment (cxnder) [inline Core/AnalysisInfo.h:106]: Punctuation

File: InfoHandler.cpp [hunk 4]
```diff
@@ -10,6 +10,7 @@
 #include "Constants.h"
 #include "CustomTypes.h"
 #include "Performance.h"
+#include "Core/TypeParser.h"
```
Comment (cxnder) [inline InfoHandler.cpp:13]: Why is this imported?

File: CustomTypes.h [hunk 5]
```diff
@@ -25,6 +25,8 @@ const std::string CFString = "CFString";
 
 const std::string MethodList = "objc_method_list_t";
 const std::string Method = "objc_method_t";
+const std::string IvarList = "objc_ivar_list_t";
```
Comment (cxnder) [inline CustomTypes.h:28]: Move this to after MethodListEntry

**Comment by cxnder** [comment]
Before -> After of this PR
image

File: Core/AnalysisInfo.h [hunk 6]
```diff
@@ -80,6 +80,38 @@ struct MethodListInfo {
     bool hasDirectSelectors() const;
 };
 
+/**
+ * A description of an Objective-C ivar.
```
Comment (jonpalmisc) [inline Core/AnalysisInfo.h:84]: Pedantic, but: "ivar" being short for "instance variable" should be conveyed at least once somewhere in this code, even though the runtime calls them "ivars". 

In general, I think "instance variable" > "ivar" in comments at least; "ivar" in names/types is fine.

This same comment applies to the next few documentation comments.

**Review by jonpalmisc — CHANGES_REQUESTED** [review]
Looks good barring a handful of small code changes. I did **not** test the code itself, I'm planning to do that once these requested changes are done.

Have you tried this on really large binaries? I'll do some testing myself as well, but whenever new features are added I like to get an idea of their performance impact.

**Comment by cxnder** [comment]
> Have you tried this on really large binaries? I'll do some testing myself as well, but whenever new features are added I like to get an idea of their performance impact.

I have not. There is a small potential for slowdown in that with the current type system (which this interacts with *much* more frequently now), we are potentially invoking the clang string type parser API up to twice as often per class as we were before. That should be mitigated when #19 is handled.

**Review by jonpalmisc — APPROVED** [review]

**Pull request merged**